client/table.go

```diff
@@ -0,0 +1,781 @@
+
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
+		var i uint64
+		b, i = roachencoding.DecodeUvarint(b)
+		v.SetUint(i)
+		return b, nil
+
+	case reflect.Float32, reflect.Float64:
+		var f float64
+		b, f = roachencoding.DecodeNumericFloat(b)
+		v.SetFloat(f)
+		return b, nil
+
+	case reflect.String:
+		var r []byte
+		b, r = roachencoding.DecodeBytes(b, nil)
+		v.SetString(string(r))
+		return b, nil
+	}
+
+	return nil, fmt.Errorf("unable to decode key: %s", v)
+}
+
+// encodePrimaryKey encodes a primary key for the table using the model object
+// v. It returns the encoded primary key.
+func (m *model) encodePrimaryKey(v reflect.Value) ([]byte, error) {
+	var key []byte
+	key = roachencoding.EncodeBytes(key, []byte(m.name))
+
+	for _, col := range m.primaryKey {
+		f, ok := m.fields[col]
+		if !ok {
+			return nil, fmt.Errorf("%s: unable to find field %s", m.name, col)
+		}
+		var err error
+		key, err = encodeTableKey(key, v.FieldByIndex(f.Index))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return key, nil
+}
+
+// decodePrimaryKey decodes a primary key for the table into the model object
+// v. It returns the remaining (undecoded) bytes.
+func (m *model) decodePrimaryKey(key []byte, v reflect.Value) ([]byte, error) {
+	var name []byte
+	key, name = roachencoding.DecodeBytes(key, nil)
+	if string(name) != m.name {
+		return nil, fmt.Errorf("%s: unexpected table name: %s", m.name, name)
+	}
+
+	for _, col := range m.primaryKey {
+		f, ok := m.fields[col]
+		if !ok {
+			return nil, fmt.Errorf("%s: unable to find field %s", m.name, col)
+		}
+		var err error
+		key, err = decodeTableKey(key, v.FieldByIndex(f.Index))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return key, nil
+}
+
+// encodeColumnKey encodes the column and appends it to primaryKey.
+func (m *model) encodeColumnKey(primaryKey []byte, column string) []byte {
+	var key []byte
+	key = append(key, primaryKey...)
+	return append(key, column...)
+}
+
+// BindModel binds the supplied interface with the named table. You must bind
+// the model for any type you wish to perform operations on. It is an error to
+// bind the same model type more than once and a single model type can only be
+// bound to a single table. The primaryKey arguments specify the columns that
+// make up the primary key.
+//
+// TODO(pmattis): Once we have a table schema we can use it to determine the
+// primary key columns.
+func (db *DB) BindModel(name string, obj interface{}, primaryKey ...string) error {
+	t := deref(reflect.TypeOf(obj))
+	if db.experimentalModels == nil {
+		db.experimentalModels = make(map[reflect.Type]*model)
+	}
+	if _, ok := db.experimentalModels[t]; ok {
+		return fmt.Errorf("%s: model '%T' already defined", name, obj)
+	}
+	fields, err := getDBFields(t)
+	if err != nil {
+		return err
+	}
+	m := &model{
+		name:       name,
+		fields:     fields,
+		primaryKey: primaryKey,
+	}
+	isPrimaryKey := make(map[string]bool)
```

Comment:
It is a set. Not quite clear what the hangup is about using `map[string]bool` in this manner. For temporary maps like this it seems quite natural.